packages/metadataform-extension/src/form.tsx

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) Jupyter Development Team.
+ * Distributed under the terms of the Modified BSD License.
+ */
+
+import { ReactWidget } from '@jupyterlab/apputils';
+import React from 'react';
+import Form, {
+  ArrayFieldTemplateProps,
+  FieldTemplateProps,
+  IChangeEvent,
+  ObjectFieldTemplateProps
+} from '@rjsf/core';
+import { JSONSchema7 } from 'json-schema';
+import {
+  PartialJSONObject,
+  PartialJSONValue,
+  ReadonlyPartialJSONObject
+} from '@lumino/coreutils';
+import { ITranslator } from '@jupyterlab/translation';
+
+import { MetadataFormWidget } from './index';
+import {
+  addIcon,
+  caretDownIcon,
+  caretUpIcon,
+  closeIcon,
+  LabIcon
+} from '@jupyterlab/ui-components';
+
+export namespace MetadataForm {
+  export interface IProperties {
+    type: string;
+    properties: { [metadataKey: string]: PartialJSONObject };
+  }
+
+  export interface IMetadataKeys {
+    [metadataKey: string]: Array<string>;
+  }
+
+  /**
+   * RJSF ui:schema.
+   */
+  export interface IUiSchema {
+    [metadataKey: string]: IUiSchemaOption;
+  }
+
+  /**
+   * RJSF ui:schema options.
+   */
+  export interface IUiSchemaOption {
+    [option: string]: any;
+  }
+
+  /**
+   * Default values
+   */
+  export interface IDefaultValues {
+    [metadataKey: string]: any;
+  }
+
+  /**
+   * Props passed to the FormWidget component
+   */
+  export interface IProps {
+    /**
+     * Properties defined from the settings.
+     */
+    properties: IProperties;
+
+    /**
+     *  Metadata keys associated to properties.
+     */
+    metadataKeys: IMetadataKeys;
+
+    /**
+     * Current data of the form.
+     */
+    formData: ReadonlyPartialJSONObject | null;
+
+    // /**
+    //  *  Errors on the form.
+    //  */
+    // errors?: {[keyError: string]: {__errors: string}};
+
+    /**
+     * Translator object
+     */
+    translator: ITranslator | null;
+
+    /**
+     * The parent object of the form.
+     */
+    parent: MetadataFormWidget;
+
+    /**
+     * The uiSchema built when loading schemas.
+     */
+    uiSchema: IUiSchema;
+
+    /**
+     * The default values for each key.
+     */
+    defaultValues: IDefaultValues;
+  }
+
+  /**
+   * Template to allow for custom buttons to re-order/remove entries in an array.
+   */
+  export const CustomArrayTemplateFactory =
+    (): React.FC<ArrayFieldTemplateProps> => {
```

Comment:
We probably would not want to have the custom templates duplicated here and in `packages/settingeditor/src/SettingsFormEditor.tsx` but instead I would prefer to move the common code to `ui-components` or a new package dedicated to customising rjsf (not sure on this idea though); there are bugs we need to fix and changes coming with the new rjsf version and having to keep the templates copies in sync manually will be cumbersome and error-prone.

Reply:
>  have the custom templates duplicated here

Agreed. Indeed, having a proper `ui-components/lib/schemaform` (_not_ imported by default) with base styling/etc. which multiple providers can subclass is clutch.

>  new rjsf version

We should take a hard look at `rjsf5`, as it might make custom templates far easier.

I would also _really_ like to see us support the data-driven `uiSchema` (#11989) directly, as the current "whole custom widget" approach is really unwieldy, and doesn't integrate particularly well with third-party tools.

Reply:
> We should take a hard look at `rjsf5`

Agreed. At least the `v4` (current stable) to include conditional fields, look like it is not available in the `v3`.